### PR TITLE
 Bug 1568503 - Ensure queued initial tasks are executed in order

### PR DESCRIPTION
**Pull request by travis79**
This ensures that tasks queued before Glean is initialized are executed in the proper order, including tasks that may have been launched after the flushing of the queue has began.

### Pull Request checklist ###

- [ ] **Quality**: This PR builds and tests run cleanly
  - `make test` runs without emitting any warnings
  - `make lint` runs without emitting any errors
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes a changelog entry or an explanation of why it does not need one
  - Any breaking changes to language binding APIs are noted explicitly

### CODE REVIEW

File: glean-core/android/src/test/java/mozilla/telemetry/glean/DispatchersTest.kt [hunk 1]
```diff
@@ -60,4 +63,50 @@ class DispatchersTest {
...
+    @Test
+    fun `queued tasks are executed in the order they are received`() {
+        var orderedString = ""
+        var curNum = 0
+
+        // Set testing mode to false as we want every call to `launch` to execute as a coroutine
+        Dispatchers.API.setTestingMode(false)
+        Dispatchers.API.setTaskQueueing(true)
+
+        // Spawn a coroutine that will add elements to the orderedString.  This will continue to add
```
Comment (badboy) [inline glean-core/android/src/test/java/mozilla/telemetry/glean/DispatchersTest.kt]: Why a string? Naturally I would have just reached for a list.

Comment (travis79) [reply glean-core/android/src/test/java/mozilla/telemetry/glean/DispatchersTest.kt]: > Did you check that the test really (reliably) fails without the changes to `Dispatchers.kt`?

I could get the new test to fail intermittently, and (locally) it appeared that the changes removed the failure.  Of course, it's still failing on CI so I may have to dig a little deeper.

Comment (travis79) [reply glean-core/android/src/test/java/mozilla/telemetry/glean/DispatchersTest.kt]: A list should work fine. I went for a string as it was easy to compare the two side by side to see when/where it went wrong, but it looks like a list gives me the same ability.

**Comment by travis79** [comment]
I modified the test to more _reliably_ fail, per your request.  In doing so I also changed the queue flushing approach to work better at executing in order.

**Comment by mdboom** [comment]
This looks great.  One thing that doesn't "feel right" is the hardcoded timeout for the initial events, just because if it is hit, results are going to be weird and we might not immediately know about it.  I wonder if it would be possible to record an error (that gets sent in the ping) if that timeout times out...

**Comment by travis79** [comment]
> This looks great. One thing that doesn't "feel right" is the hardcoded timeout for the initial events, just because if it is hit, results are going to be weird and we might not immediately know about it. I wonder if it would be possible to record an error (that gets sent in the ping) if that timeout times out...

Yeah, I understand.  I'm not a fan of timeouts either but in this case it felt better than having a situation where Glean would hang.  It wasn't necessary before because it wasn't using the `runBlocking` to ensure that the queued items get executed before any items that come in after the queue, so I added it as a safety measure.

I don't think that either `invalid_value` or `invalid_label` really fit with a timeout error, what do you think about filing a bug about adding an error for this case as follow-up work?

**Comment by mdboom** [comment]
Yeah -- sounds good to me to file a bug to come back to this.  Thanks!

**Comment by mdboom** [comment]
Oh, and I'll file that bug...

File: glean-core/android/src/main/java/mozilla/telemetry/glean/Dispatchers.kt [hunk 2]
```diff
@@ -111,13 +116,28 @@ internal object Dispatchers {
...
+                }
+                queueInitialTasks = false
+                taskQueue.clear()
+            }
+
+            // In order to ensure that the queued tasks are executed in the proper order, we will
+            // wait up to 5 seconds for it to complete, otherwise we will reset the flag so that
+            // new tasks may continue to run.
+            runBlocking {
+                val result = withTimeoutOrNull(QUEUE_PROCESSING_TIMEOUT_MS) {
```
Comment (badboy) [inline glean-core/android/src/main/java/mozilla/telemetry/glean/Dispatchers.kt]: What is the return value of this block when it doesn't time out?

File: glean-core/android/src/test/java/mozilla/telemetry/glean/DispatchersTest.kt [hunk 3]
```diff
@@ -60,4 +65,52 @@ class DispatchersTest {
         assertEquals("Tasks have executed", 3, threadCanary)
         assertEquals("Task queue is cleared", 0, Dispatchers.API.taskQueue.size)
     }
+
+    @Test
+    fun `queued tasks are executed in the order they are received`() {
+        val orderedList = mutableListOf<String>()
```
Comment (badboy) [inline glean-core/android/src/test/java/mozilla/telemetry/glean/DispatchersTest.kt]: Well, I guess I would have gone for a list of ints, but converting them to a string is just as fine

File: glean-core/android/src/test/java/mozilla/telemetry/glean/DispatchersTest.kt [hunk 4]
```diff
@@ -60,4 +65,52 @@ class DispatchersTest {
...
+        Dispatchers.API.setTaskQueueing(true)
+
+        // Spawn a coroutine that will add elements to the orderedList.  This will continue to add
+        // elements to the queue until there are at least 50 elements in the queue. At that point,
+        // the second coroutine below will flush and disable the queue.
+        val addJob = GlobalScope.launch {
+            (0..100).forEach { num ->
+                Dispatchers.API.launch {
+                    orderedList.add("$num")
+                }.also { job ->
```
Comment (badboy) [inline glean-core/android/src/test/java/mozilla/telemetry/glean/DispatchersTest.kt]: Oh noes, another operator

File: glean-core/android/src/main/java/mozilla/telemetry/glean/Dispatchers.kt [hunk 5]
```diff
@@ -111,13 +116,28 @@ internal object Dispatchers {
...
+                    job.join()
+                }
+
+                if (result == null) {
```
Comment (badboy) [inline glean-core/android/src/main/java/mozilla/telemetry/glean/Dispatchers.kt]: In case of an error it could be that tasks stick around in the `taskQueue`, right?

**Comment by badboy** [comment]
Stable test is failing. Looks like we're getting OOM-killed again.
Is this rebased on latest master?

I might have an idea how to reduce the impact of that...

**Comment by badboy** [comment]
Oh wait, the oom mitigation wasn't landed on master

File: glean-core/android/src/main/java/mozilla/telemetry/glean/Dispatchers.kt (hunk 2 above)
Comment (travis79) [reply glean-core/android/src/main/java/mozilla/telemetry/glean/Dispatchers.kt]: It returns the coroutine back.  See [here](https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/with-timeout-or-null.html)

File: glean-core/android/src/test/java/mozilla/telemetry/glean/DispatchersTest.kt (hunk 3 above)
Comment (travis79) [reply glean-core/android/src/test/java/mozilla/telemetry/glean/DispatchersTest.kt]: Probably lighter weight yet to go with Ints, point taken.

File: glean-core/android/src/test/java/mozilla/telemetry/glean/DispatchersTest.kt (hunk 4 above)
Comment (travis79) [reply glean-core/android/src/test/java/mozilla/telemetry/glean/DispatchersTest.kt]: While my original Kotlinism works, I could be more clear.  Let me rephrase it...

File: glean-core/android/src/main/java/mozilla/telemetry/glean/Dispatchers.kt (hunk 5 above)
Comment (travis79) [reply glean-core/android/src/main/java/mozilla/telemetry/glean/Dispatchers.kt]: Yes, good catch!  The queue should probably be emptied if we error.

**Comment by travis79** [comment]
> This LGTM, thanks for tackling it. Can we make sure this test is not flaky (i.e. execute it 4-5 times on CI before merging?). Moreover, does the test consistently fail without your changes?

Thanks for reviewing!  Yes, the test fails consistently now.  That was one of the things that @badboy brought up in his review and so I made sure it was maximum chance to fail.  And as far as I can tell, it's not flaky just had some issues with the CI container size causing failures last week.
